server/src/service/service.rs

```diff
@@ -0,0 +1,210 @@
+            if attributes.get("locked").is_some() {
+                // this if condition is probably wrong
+                // how to access &AttributeValue value
+                locked.push(item)
+            } else {
+                unlocked.push(item)
+            }
+        }
+
+        Ok((unlocked, locked))
+    }
+
+    #[dbus_interface(out_args("unlocked", "prompt"))]
+    pub async fn unlock(
+        &mut self,
+        objects: Vec<OwnedObjectPath>,
+    ) -> fdo::Result<(Vec<OwnedObjectPath>, OwnedObjectPath)> {
+        // manage unlock state in memory
+        // when do we need to prompt?
+        let mut unlocked: Vec<OwnedObjectPath> = Vec::new();
+
+        'main: for object in objects {
+            for collection in &mut self.collections {
+                if collection.path().await.as_str() == object.as_str() {
+                    if collection.locked().await {
+                        collection.set_locked(false).await;
+                        unlocked.push(object.clone());
+                    } else {
+                        break 'main;
+                    }
+                }
+            }
+        }
+
+        if unlocked.len() == 0 {
+            unlocked.push(ObjectPath::default().into());
+        }
+
+        let prompt = ObjectPath::default().into(); // temporarily
+
+        Ok((unlocked, prompt))
+    }
+
+    #[dbus_interface(out_args("locked", "prompt"))]
+    pub async fn lock(
+        &mut self,
+        objects: Vec<OwnedObjectPath>,
+    ) -> fdo::Result<(Vec<OwnedObjectPath>, OwnedObjectPath)> {
+        // manage lock state in memory
+        // when do we need to prompt?
+        let mut locked: Vec<OwnedObjectPath> = Vec::new();
+
+        for object in objects {
+            for collection in &mut self.collections {
+                if collection.path().await.as_str() == object.as_str() {
+                    if !collection.locked().await {
+                        collection.set_locked(true).await;
+                        locked.push(object.clone());
+                    }
+                }
+            }
+        }
+
+        if locked.len() == 0 {
+            locked.push(ObjectPath::default().into());
+        }
+
+        let prompt = ObjectPath::default().into(); // temporarily
+
+        Ok((locked, prompt))
+    }
+
+    pub async fn get_secrets(&self, items: Vec<OwnedObjectPath>) {
+        // WIP: not complete
+        if KEYRING.get().unwrap().n_items().await == 0 {
+            return;
+        }
+
+        // call item iface get_secret multiple times ?
+
+        for in_collection in &self.collections {
+            for item in &items {
+                if in_collection.path().await.as_str() == item.as_str() {}
+            }
+        }
+    }
+
+    pub async fn read_alias(&self, name: &str) -> ObjectPath {
+        let mut ret = ObjectPath::default();
+        for collection in &self.collections {
+            if collection.label().await == name {
```

Comment:
this is the alias of the collection, not it is label

Reply:
updated the entire method.